src/metabase/api/cloud_migration.clj

```diff
@@ -0,0 +1,245 @@
+
+(defn- progress-file-input-stream
+  "File input stream that calls on-percent-progress with current read progress as int from 0 to 100.
+  Does not call on-percent-progress for the same value twice. "
+  [^File file on-percent-progress]
+  (let [input-stream             (io/input-stream file)
+        length                   (.length file)
+        *bytes                   (atom 0)
+        on-percent-progress-memo (memoize on-percent-progress) ;; memoize so we don't repeat calls
+        add-bytes                #(on-percent-progress-memo (int (* 100 (/ (swap! *bytes + %) length))))
+        f                        (fn [ret]
+                                   (cond
+                                     ;; -1 is end of stream
+                                     (= -1 ret)  (on-percent-progress-memo 100)
+                                     (char? ret) (add-bytes 1)
+                                     (int? ret)  (add-bytes ret))
+                                   ret)]
+    (proxy [InputStream] []
+      (read
+       ([]
+        (f (.read input-stream)))
+        ([^bytes b]
+         (f (.read input-stream b)))
+        ([^bytes b off len]
+         (f (.read input-stream b off len))))
+      (close [] (.close input-stream)))))
+
+(defn- set-progress
+  "Attempt to set id to state and progress.
+  Throws if the migration has already reached a terminal state (e.g. cancelled).
+  This is the main cluster coordination mechanism for migrations, since any instance
+  can cancel the migration, not just the one that initiated it."
+  [id state progress]
+  (when (= 0 (t2/update! CloudMigration :id id :state [:not-in terminal-states]
+                         {:state state :progress progress}))
+    (throw (ex-info "Cannot update migration in terminal state" {:terminal true}))))
+
+(defn- migrate!
+  "Migrate this instance to Metabase Cloud.
+  Will exit early if migration has been cancelled in any cluster instance."
+  [{:keys [id upload_url]} & {:keys [retry?]}]
+  ;; dump-to-h2 starts behaving oddly if you try to dump repeatly to the same file
+  ;; in the same process.
+  (let [dump-file (io/file (str "cloud_migration_dump_" (random-uuid) ".mv.db"))
+        ;; Note: this will still set progress up to twice for each percent
+        ;; because e.g. 70 and 71 upload percent both map to 50+35 migration percent.
+        on-upload-progress #(set-progress id :upload (int (+ 50 (* 50 (/ % 100)))))]
+    (try
+      (when retry?
+        (t2/update! CloudMigration :id id {:state :init}))
+
+      (log/info "Setting read-only mode")
+      (set-progress id :setup 1)
+      (read-only-mode! true)
+      (when (cluster?)
+        (log/info "Cluster detected, waiting for read-only mode to propagate")
+        (Thread/sleep (int (* 1.5 setting.cache/cache-update-check-interval-ms))))
+
+      (log/info "Dumping h2 backup to" (.getAbsolutePath dump-file))
+      (set-progress id :dump 20)
+      (binding [*ignore-read-only-mode* true]
```

Comment:
`*ignore-read-only-mode*` is not doing what you think it does here.

all toucan2 operations in `dump-to-h2/dump-to-h2!` are working directly with the table, not the models. so I think if you remove this binding, it'll still work.

See https://github.com/metabase/metabase/pull/41317/files#r1586163568

Reply:
Idea: I think you should make `read-only-mode-exceptions` dynamic and dissoc it with models that `dump-to-h2/dump-to-h2!` needs to update during its execution.

Reply:
Spoke a little bit too soon. The CI tests show the error mode that originally led me to add this var. You can see it [here](https://github.com/metabase/metabase/actions/runs/8910154054/job/24468847652?pr=41317#step:3:1112): a readonly error during key rotation.

Turns out this error doesn't always show up. If you start a repl, call `(dev)`, and run the unit tests without the var, it doesn't show up. But if you call `(do (dev) (start!))` and then call the tests it does show up. I'm not sure exactly how CI is doing it but it does show up there too.

Added it back then.

Reply:
Did you remove this after you changed `read-only-mode-exceptions` to a list of table names? if so then it makes sense why the flag works.

wdyt about my idea of making `read-only-mode-exceptions` a dynamic var? then we can remove `*ignore-read-only-mode*`

Reply:
I had this error before changing the exception to table names, so I don't think that was it.

Regarding making `read-only-mode-exceptions` dynamic... I think that's an option, but I don't feel very inclined to go explore it. It doesn't seem to be an obvious improvement, it's more of a side grade, so it doesn't feel like a great use of my time. I'd have to go change it, then debug it, etc, and I'd end up at the same place.

The core problem is around how the dump key rotation works by binding the current `mdb.connection/*application-db*` to edit the h2 dump. That's why the read-only mode is triggered during dump. But I'm not looking to go fixing dump-to-h2 more than I have to for this feature.